### PR TITLE
IGNITE-22505 Reuse pages where possible

### PR DESCRIPTION
**Pull request by Phillippko**
https://issues.apache.org/jira/browse/IGNITE-22505

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryMvTableStorageTest.java [hunk 1]
```diff
@@ -57,10 +60,18 @@ public class VolatilePageMemoryMvTableStorageTest extends AbstractMvTableStorage
 
     @BeforeEach
     void setUp(
+            TestInfo testInfo,
             @InjectConfiguration VolatilePageMemoryStorageEngineConfiguration engineConfig,
             @InjectConfiguration("mock.profiles.default = {engine = \"aimem\"}")
             StorageConfiguration storageConfiguration
     ) {
+        if (testInfo.getTags().contains(LOW_MEMORY_TAG)) {
```
Comment (sashapolo) [inline modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryMvTableStorageTest.java:68]: Where did this come from?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/IndexStorageFactory.java [hunk 2]
```diff
@@ -274,7 +274,7 @@ private <T> IndexTreeAndMeta<T> createIndexTree(StorageIndexDescriptor descripto
         try {
             PageMemory pageMemory = tableStorage.dataRegion().pageMemory();
 
-            long metaPageId = pageMemory.allocatePage(tableStorage.getTableId(), partitionId, PageIdAllocator.FLAG_AUX);
+            long metaPageId = pageMemory.allocatePageNoReuse(tableStorage.getTableId(), partitionId, PageIdAllocator.FLAG_AUX);
```
Comment (sashapolo) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/IndexStorageFactory.java:277]: Why don't we reuse pages here?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java [hunk 3]
```diff
@@ -164,7 +164,7 @@ private FreeListImpl createFreeList(
             boolean initNew = false;
 
             if (meta.freeListRootPageId() == 0) {
-                long rootPageId = pageMemory.allocatePage(getTableId(), partId, FLAG_AUX);
+                long rootPageId = pageMemory.allocatePageNoReuse(getTableId(), partId, FLAG_AUX);
```
Comment (sashapolo) [inline modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:167]: Why no reuse here?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/PageIdAllocator.java [hunk 4]
```diff
@@ -39,14 +48,85 @@ public interface PageIdAllocator {
...
+            @Nullable ReuseList reuseList,
+            @Nullable ReuseBag bag,
+            boolean useRecycled,
+            int groupId,
+            int partitionId,
+            byte flags
+    ) throws IgniteInternalCheckedException {
+        long pageId = 0;
+
+        if (reuseList != null) {
```
Comment (sashapolo) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/PageIdAllocator.java:93]: Why did we start ignoring the `useRecycled` flag?

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 3 above)
Comment (Phillippko) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:167]: We use free lists to reuse pages, so when creating free list itself we have nothing to reuse pages from. Or I'm wrong?

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryMvTableStorageTest.java (hunk 1 above)
Comment (Phillippko) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryMvTableStorageTest.java:68]: By default we allocate 3 gigabytes of memory, so we need a lot of iterations to overflow memory if we have leakages. I.e. without my fix volatile memory fails only after 51 thousand iterations with default memory settings

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 3 above)
Comment (Phillippko) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:167]: So starting persistent partiton without previous meta we actually don't reuse any pages. Free list is just created, it is empty and we have to allocate new pages for all structures. Don't know what to do with that

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryMvTableStorageTest.java (hunk 1 above)
Comment (Phillippko) [reply modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/VolatilePageMemoryMvTableStorageTest.java]: Removed. Moved test to volatile test class (at first didn't realize that it's not valid for persistent), and added more iterations - it still takes only 4 seconds, but without my fix overflows even with default memory settings

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java (hunk 3 above)
Comment (sashapolo) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryTableStorage.java:167]: We have both a `FreeList` and a `ReuseList`, I always confuse them. I thought we used `ReuseList` for empty pages

File: modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryMvTableStorageTest.java [hunk 5]
```diff
@@ -57,6 +58,7 @@ public class PersistentPageMemoryMvTableStorageTest extends AbstractMvTableStora
 
     @BeforeEach
     void setUp(
+            TestInfo testInfo,
```
Comment (sashapolo) [inline modules/storage-page-memory/src/test/java/org/apache/ignite/internal/storage/pagememory/PersistentPageMemoryMvTableStorageTest.java:61]: ?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/PageIdAllocator.java [hunk 6]
```diff
@@ -39,14 +48,85 @@ public interface PageIdAllocator {
...
+        assert pageId != 0;
+
+        assert partitionId(pageId) >= 0 && partitionId(pageId) <= MAX_PARTITION_ID : toDetailString(pageId);
+
+        assert flag(pageId) != FLAG_DATA || itemId(pageId) == 0 : toDetailString(pageId);
+
+        return pageId;
+    }
+
+    private static long replacePartitionId(long pageId, int partId) {
```
Comment (sashapolo) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/PageIdAllocator.java:123]: Why did you remove the javadoc?

**Review by sashapolo — APPROVED** [review]
